### PR TITLE
CC-Widgets Packages setup

### PR DESCRIPTION
**Pull request by Shreyas281299**
## Changes
Adds basic setup for widget
### Functionalities
- Adds ability to build providers - yarn build:providers
- Adds ability to build widgets - yarn build:widgets
- Adds ability to build both at once (providers should be built first) yarn run build
- Adds ability to build samples, both react based samples and web-component- yarn run build:samples
- Adds ability to run react samples- yarn samples:serve-react
- Adds ability to run web-component samples - yarn samples:serve-wc
- Adds ability to run tests(tests are only setup in widget-cc-station-login)

### Testing
- Built providers
- Build widgets
- Build samples
- Ran samples
- Ran test for widget-cc-station-login

### CODE REVIEW

File: .babelrc.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
```
Comment (Shreyas281299) [inline .babelrc.js:1]: Created common babelrc which will be imported in each workspace. babelrc is required in each workspace to run jest tests

File: .gitignore [hunk 2]
```diff
@@ -43,6 +43,7 @@ reports/
 !.env.default
 !package.json
 !package-lock.json
+!tsconfig.json
```
Comment (Shreyas281299) [inline .gitignore:46]: tsconfig was getting ignore, was causing issues.

File: docs/react-samples/package.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (Shreyas281299) [inline docs/react-samples/package.json:1]: The sample app is also a workspace. This will allow us to easily link updated widgets locally.

Comment (Shreyas281299) [reply docs/react-samples/package.json:1]: All the files in this are demo, these will be updated when we add the actual implementation of these widgets. 
We can still run and see the basic providers and widgets working and sharing states.

File: docs/react-samples/src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
```
Comment (Shreyas281299) [inline docs/react-samples/src/App.tsx:1]: Dummy code for a simple demo.

File: docs/react-samples/webpack.config.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
+const path = require('path');
+const HtmlWebpackPlugin = require('html-webpack-plugin');
+const {merge} = require('webpack-merge');
+
+const baseConfig = require('../../webpack.config');
```
Comment (Shreyas281299) [inline docs/react-samples/webpack.config.js:5]: Using base webpack config here also

File: docs/web-component-samples/index.html [hunk 6]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
```
Comment (Shreyas281299) [inline docs/web-component-samples/index.html:1]: This is a sample using web-components and NO react. Again dummy code to just show that widgets can be used as web-components

File: packages/@webex/widget-cc-station-login/.babelrc.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+const baseConfig = require('../../../.babelrc');
```
Comment (Shreyas281299) [inline packages/@webex/widget-cc-station-login/.babelrc.js:1]: using base babelrc config.

**Review by arun3528 — CHANGES_REQUESTED** [review]
I dont think you have to follow the structure of packages/@webex structure here 

yarn workspace can just relay on the folder structure like widgets/cc-station-login etc 

the npm package where you are going to publish can be @webex/cc-station-login

The naming looks little messed up i know we did not follow this on the sdk side

File: .gitignore [hunk 8]
```diff
@@ -43,6 +43,7 @@ reports/
...
 node_modules/
```
Comment (arun3528) [inline .gitignore:47]: this should be /node_moidules/ once you add yarn i guess

File: packages/@webex/widget-cc-station-login/src/hooks.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+export const useStationLogin = () => {
+  return {name: 'StationLogin'};
```
Comment (arun3528) [inline packages/@webex/widget-cc-station-login/src/hooks.ts:2]: can we confirm the naming of this file as hooks if its going to be just a helper file for components should be name it utils or something as it gets confused with react hooks

if we going to import react stuff here i assume its ok

File: packages/@webex/widget-cc-station-login/src/station-login/index.tsx [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+import store from '@webex/widget-provider';
+import {observer} from 'mobx-react';
+
+import StationLoginPresentational from './station-login.presentational';
+import {useStationLogin} from '../hooks';
+
+const StationLogin: React.FunctionComponent = observer(() => {
+  const {loginState, setLoginState, ccSdk, isAvailable} = store;
+  const result = useStationLogin();
+
```
Comment (arun3528) [inline packages/@webex/widget-cc-station-login/src/station-login/index.tsx:12]: we will need to add some documentation here

File: packages/@webex/widget-provider/package.json [hunk 11]
```diff
@@ -0,0 +1,42 @@
+{
```
Comment (arun3528) [inline packages/@webex/widget-provider/package.json:1]: should we rename this to something else , provider is too generic , should it be called widgets store or something ?

File: packages/@webex/widget-provider/src/store.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
+import {makeAutoObservable, observable} from 'mobx';
```
Comment (arun3528) [inline packages/@webex/widget-provider/src/store.ts:1]: each substore are logical chunks lets name them based on features 

AgentStore , taskStore etc 
 we can have a call to talk about this

Comment (mkesavan13) [reply packages/@webex/widget-provider/src/store.ts:1]: Hi @arun3528 

I agree. We would need to do that, except that we won't do it right away. At the start, we will keep it limited to just one common store and later when we build the Tasks widget, we will introduce a task store as part of that PR, agent store as part of that PR and so on, although for AgentLogin, we need not have a separate store. Just the common store shall be good to consume. 

Do we want to keep this a non-blocking comment and address it as part of future PRs?

File: package.json [hunk 13]
```diff
@@ -5,10 +5,24 @@
...
+    "webpack-cli": "^5.1.4",
+    "webpack-dev-server": "^5.1.0"
+  },
+  "scripts": {
+    "build": "yarn run build:providers && yarn run build:widgets",
+    "build:providers": "yarn workspace @webex/widget-provider build:src",
+    "build:widgets": "yarn workspace @webex/widget-cc-station-login build:src && yarn workspace @webex/widget-cc-user-state build:src",
+    "samples:build": "yarn workspace @webex/react-samples-app build:src && yarn workspace @webex/web-component-samples-app build:src",
+    "samples:serve-react": "yarn workspace @webex/react-samples-app serve",
+    "samples:serve-wc": "yarn workspace @webex/web-component-samples-app serve"
```
Comment (mkesavan13) [inline package.json:26]: Non-blocking: Can we have one serve command that shows both wc as well as react apps?

File: packages/@webex/widget-cc-station-login/src/station-login/index.tsx (hunk 10 above)
Comment (mkesavan13) [reply packages/@webex/widget-cc-station-login/src/station-login/index.tsx:12]: What kind of documentation are you looking for Arun? Just code comments or something like API reference?

File: packages/@webex/widget-cc-user-state/src/user-state/user-state.presentational.tsx [hunk 14]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+
+import {IUserState} from './use-state.types';
+
+const UserStatePresentational: React.FunctionComponent<IUserState> = (props) => {
+  return (
+    <>
+      <h1 data-testid="user-state-heading">{props.name}</h1>
```
Comment (mkesavan13) [inline packages/@webex/widget-cc-user-state/src/user-state/user-state.presentational.tsx:8]: Why do we have this `data-testid` everywhere? 

Are we also setup for the UI testing?

File: packages/@webex/widget-provider/package.json (hunk 11 above)
Comment (mkesavan13) [reply packages/@webex/widget-provider/package.json:1]: `@webex/widgets-store` sounds good to me

File: packages/@webex/widgets/package.json [hunk 15]
```diff
@@ -33,8 +33,8 @@
   "dependencies": {
     "@webex/component-adapter-interfaces": "^1.30.5",
     "@webex/components": "1.275.2",
-    "@webex/sdk-component-adapter": "1.112.12",
-    "webex": "2.60.4"
+    "@webex/sdk-component-adapter": "1.112.11",
+    "webex": "2.60.2"
```
Comment (mkesavan13) [inline packages/@webex/widgets/package.json:37]: This is resetting Parimala's change. Probably a conflict resolution issue. Please check

**Review by mkesavan13 — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 8 above)
Comment (Shreyas281299) [reply .gitignore:47]: node_modules/ is what is ignorig the node_modules inside each workspace. If I change it to /node_modules/ then its not ignoring node_modules.

File: packages/@webex/widget-cc-station-login/src/hooks.ts (hunk 9 above)
Comment (Shreyas281299) [reply packages/@webex/widget-cc-station-login/src/hooks.ts:2]: Changed to helper.ts

File: packages/@webex/widget-provider/package.json (hunk 11 above)
Comment (Shreyas281299) [reply packages/@webex/widget-provider/package.json:1]: Changed to widgets-store

File: packages/@webex/widget-cc-user-state/src/user-state/user-state.presentational.tsx (hunk 14 above)
Comment (Shreyas281299) [reply packages/@webex/widget-cc-user-state/src/user-state/user-state.presentational.tsx:8]: This I've added for simple unit testing for UI. This is used in the test that checks if the correct prop is rendered. I think that much testing for a basic UI is okay? What do you think?

File: packages/@webex/widgets/package.json (hunk 15 above)
Comment (Shreyas281299) [reply packages/@webex/widgets/package.json:37]: I missed this. Reverted this change

File: packages/@webex/widget-cc-station-login/src/hooks.ts
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply packages/@webex/widget-cc-station-login/src/hooks.ts:1]: Screenshot 2024-11-11 at 11 02 44 PM
The idea behind moving the hooks/helper file outside is that a single widget can have multiple conponents and all these components will use only 1 hooks/helper file. Benefit for this is that we can easily ship the hooks file separately that will provide the business logic and the presentational layer is present inside the folders.
@mkesavan13  please add anything that Ive missed.

File: docs/web-component-samples/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,10 @@
+import {WebStationLogin} from '@webex/widget-cc-station-login';
+import {WebUserState} from '@webex/widget-cc-user-state';
+
+if (!customElements.get('web-user-state')) {
```
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply docs/web-component-samples/src/index.ts:4]: Moved this logic inside each widget- now developers only need to import the package and then the custom web components will be available

File: package.json (hunk 13 above)
Comment (Shreyas281299) [reply package.json:27]: https://github.com/user-attachments/assets/1c988e62-719e-4274-b039-c6a21f5dcfd4

This is kind of hacky

What I'm doing?
- Running the react-samples-app in the background (this is at localhost:300)
- Opening the home page 

What happens when we click on the buttons?
- When we click on "Contact Center in a react app", I'm re-directly to localhost:3000
- When we click on "Contact Center widgets as web components" - Im opening the html file that contains the web-component

Problem 
- Not sure if we will be able to run this in github pages
- Not sure if this is a correct way of running both the samples app from the same page

Comment (Shreyas281299) [reply package.json:27]: Now that I think about it maybe we should have another application that imports both react and web-component app and then just renders them. And we can run this app using webpack-dev-server

**Review by mkesavan13 — CHANGES_REQUESTED** [review]

File: packages/@webex/widget-cc-station-login/tests/station-login/station-login.hooks.ts [hunk 17]
```diff
@@ -0,0 +1,8 @@
+import {useStationLogin} from '../../src/helper';
+
+describe('Station login hooks', () => {
```
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply packages/@webex/widget-cc-station-login/tests/station-login/station-login.hooks.ts:3]: Changed

File: packages/@webex/widget-cc-station-login/tsconfig.json [hunk 18]
```diff
@@ -0,0 +1,11 @@
+{
+    "extends": "../../../tsconfig.json",
+    "compilerOptions": {
+      "outDir": "./dist",
+      "declaration": true,
+      "declarationDir": "./dist/types"
+    },
+    "include": [
+      "./src"
+    ]
```
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply packages/@webex/widget-cc-station-login/tsconfig.json:10]: Moved to global config

File: packages/@webex/widget-cc-station-login/webpack.config.js [hunk 19]
```diff
@@ -0,0 +1,12 @@
...
+const path = require('path');
+
+const baseConfig = require('../../../webpack.config');
+
+module.exports = merge(baseConfig, {
+  output: {
+    path: path.resolve(__dirname, 'dist'),
+    filename: 'index.js', // Set the output filename to index.js
+    libraryTarget: 'commonjs2',
+  },
```
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply packages/@webex/widget-cc-station-login/webpack.config.js:11]: moved to global config

File: packages/@webex/widgets-store/package.json [hunk 20]
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "@webex/widgets-store",
+  "description": "Webex Contact Center Widgets: Widgets Provider",
```
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply packages/@webex/widgets-store/package.json:3]: Updated

**Comment by Shreyas281299** [comment]
> I dont think you have to follow the structure of packages/@webex structure here
> 
> yarn workspace can just relay on the folder structure like widgets/cc-station-login etc
> 
> the npm package where you are going to publish can be @webex/cc-station-login
> 
> The naming looks little messed up i know we did not follow this on the sdk side

Moved the cc-widgets out of @webex folder.

File: packages/@webex/widget-cc-station-login/src/station-login/index.tsx (hunk 10 above)
Comment (arun3528) [reply packages/@webex/widget-cc-station-login/src/station-login/index.tsx]: code comments in general

File: packages/@webex/widget-provider/src/store.ts (hunk 12 above)
Comment (arun3528) [reply packages/@webex/widget-provider/src/store.ts]: sounds good to me

**Comment by arun3528** [comment]
@Shreyas281299 @mkesavan13  i am still not sure about the folder structure and the naming we have to make it little meaningful 

one thing is like naming as its in widgets folder and repo we dont have to add workd widget-cc word all the time could be just cc-stationLogin

**Comment by Shreyas281299** [comment]
> Screenshot 2024-11-12 at 10 27 11 AM @Shreyas281299 @mkesavan13 i am still not sure about the folder structure and the naming we have to make it little meaningful
> 
> one thing is like naming as its in widgets folder and repo we dont have to add workd widget-cc word all the time could be just cc-stationLogin
> 
> you can also change the @webex/widget folder as well if needed

- Moved widgets inside a contact-center folder
- renamed all the widgets to 'cc-X'

File: packages/@webex/widget-cc-station-login/src/station-login/index.tsx (hunk 10 above)
Comment (mkesavan13) [reply packages/@webex/widget-cc-station-login/src/station-login/index.tsx]: This is just dummy code. We will add comments and documentation when we write the actual code

File: packages/@webex/widget-provider/package.json (hunk 11 above)
Comment (mkesavan13) [reply packages/@webex/widget-provider/package.json]: This is now changed to `@webex/cc-store` as a final call

File: packages/@webex/widget-cc-station-login/src/station-login/index.tsx [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+    isAvailable,
+  };
+  return <StationLoginPresentational {...props} />;
+});
+
+const WebStationLogin = r2wc(StationLogin);
+
+if (!customElements.get('widget-cc-station-login')) {
+  customElements.define('widget-cc-station-login', WebStationLogin);
+}
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply packages/@webex/widget-cc-station-login/src/station-login/index.tsx]: We will do this as part of [SPARK-580563](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-580563)

**Review by mkesavan13 — APPROVED** [review]

**Pull request merged**